File: 0663-equal-tree-partition/0663-equal-tree-partition.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def checkEqualTree(self, root: Optional[TreeNode]) -> bool:
        def dfs(node: Optional[TreeNode]) -> int: 
            if not node:
                return 0
            
            seen.append(dfs(node.left) + dfs(node.right) + node.val)
            return seen[-1]
        
        seen = []
        total = dfs(root)
        seen.pop()
        return total / 2.0 in seen

File: 0663-equal-tree-partition/0663-equal-tree-partition.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
private:
    int dfs(TreeNode* node, vector<int>& seen){
        if (!node) return 0;
        seen.push_back(dfs(node -> left, seen) + dfs(node -> right, seen) + node -> val);
        return seen.back();
    }

public:
    bool checkEqualTree(TreeNode* root) {
        vector<int> seen;
        int total = dfs(root, seen);
        seen.pop_back();
        return find(seen.begin(), seen.end(), total / 2.0) != seen.end();
    }
};